### PR TITLE
[feat] Support quantization before weight resharding

### PR DESCRIPTION
**Pull request by hy2826**
# What does this PR do ?

Support quantization before weight resharding.

# Changelog

- Add specific line by line info of high level changes in this PR.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ No ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

### **Motivation**
In many current RL framework implementation such as VeRL, if we want to use low-bit quantization in rollout, the synchronization of weights between the Trainer worker and the Rollout worker follows a "Gather-then-Quantize" pattern.

**The Current Workflow:**
Gather: High-precision weights ($BF16/FP16$) are collected from different TP/PP ranks via all_gather or broadcast.
Quantize: The target inference rank receives the full high-precision weight and then performs quantization locally.

image

This may create communication overhead: for example, transferring $BF16$ data requires $2\times$ the bandwidth compared to $FP8$. As model sizes grow, this sync becomes a significant latency floor for each rollout iteration.

### Our Design
For blockwise quantization, we propose shifting the quantization responsibility to the source ranks. By quantizing the weight shards locally before they enter the communication collective, we can reduce the data volume by $50%$ for FP8 quantization.

**Technical Workflow**
Take FP8 quantization as an example, the proposed "Quantize-then-Gather" approach involves three steps:
Local Quantization: Each Rank $i$ takes its local shard $W_i$ (in $BF16$) and computes: 
$W_{i, fp8}$: The quantized shard.
$S_i$: The corresponding scaling factor (Scale).
Low-Precision Communication: Perform all_gather on the $FP8$ tensors and the associated Scales. Since $FP8$ is 1-byte, the communication volume is halved.
Shard Assembly: The target rank directly concatenates the $FP8$ shards. Since the quantization was done per-shard, the metadata (scales) are managed alongside the data to ensure numerical consistency.

image

### Mathematical Representation
Instead of:
$$ W_{global} = \text{Gather}(W_0, W_1, \dots, W_n) \implies W_{fp8} = \text{Quantize}(W_{global}) $$
We implement:
$$ [W_{i, fp8}, S_i] = \text{Quantize}(W_i) \implies W_{fp8} = \text{Gather}(W_{0, fp8}, \dots, W_{n, fp8}) $$

Thus, the result of our design and the original pipeline should be bitwise-equal. 

### Usage
We introduce a new function named export_hf_weights_quant in auto_bridge.py and modify subsequent dependencies to support this new function. Compared with export_hf_weights, it takes three additional arguments: 

1. quant_block_size: a tuple stating the quantization block size, such as (1,32).
2. should_quantize: a function that takes a string that is the Megatron weight name and decides whether or not to quantize the corresponding weight.
3. quant_fn: a function to quantize the weight, and takes an additional optional parameter of quantization block size.

### Experiment Results: reduced wall-clock time
For Qwen3-30B-A3B with FP8 quantization, on 8 * H100(80GB), after one step of warm-up, our design reduced the wall-clock time from 22.103s to 14.429s, achieving a reduction of 34.7% compared with the original "gather-then-quantize" pipeline. 

## Summary by CodeRabbit

* **New Features**
  * Added quantized weight export workflow from Megatron to HuggingFace format, including support for custom quantization functions and block size configuration.
  * Extended export capabilities with quantization parameters while maintaining backward compatibility with existing non-quantized export paths.

### CODE REVIEW

File: src/megatron/bridge/models/conversion/auto_bridge.py [hunk 1]
```diff
@@ -15,14 +15,12 @@
 import dataclasses
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/auto_bridge.py:17]: needs tests coverage

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 2]
```diff
@@ -1017,6 +724,79 @@ def stream_weights_megatron_to_hf(
                     # Regular case - yield the tensor normally
                     yield HFWeightTuple(hf_name, final_tensor)
 
+    def stream_weights_megatron_to_hf_quant(
+        self,
+        megatron_model: Union[MegatronModel, List[MegatronModel]],
+        hf_pretrained: HFPreTrained,
+        should_quantize: callable,
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py]: why should quantize is a callable?

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 3]
```diff
@@ -1017,6 +724,79 @@ def stream_weights_megatron_to_hf(
...
+            if merge_adapter_weights and "to_wrap.weight" in task.global_param_name:
+                task_global_base_prefix, _, _ = task.global_param_name.partition(".to_wrap.weight")
+                adapter_tasks = adapter_tasks_by_base.get(task_global_base_prefix)
+            if merge_adapter_weights and adapter_tasks:
+                adapter_weights = self.materialize_adapter_weights(adapter_tasks)
+                converted_weights_dict = self._merge_lora_adapter_weights(
+                    megatron_model,
+                    converted_weights_dict,
+                    adapter_weights,
+                )
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py]: plz check code rabbit, if not relevant, just resolve

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 4]
```diff
@@ -1449,6 +1227,35 @@ def _megatron_to_hf_registered_impl(
             merge_adapter_weights=merge_adapter_weights,
         )
 
+    @stream_weights_megatron_to_hf_quant.impl((source, target))
```
Comment (yaoyu-33) [inline src/megatron/bridge/models/conversion/model_bridge.py:2063]: possible to make a quant bridge? Model bridge can inherit quant bridge similar to peft bridge. Easier to maintain code. You can move all quant methods there.

**Comment by yaoyu-33** [comment]
@hy2826 please check comment

**Comment by hy2826** [comment]
@yaoyu-33 Thanks a lot for the comments! We will make modification to the codes and push the changes soon.

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 2 above)
Comment (hy2826) [reply src/megatron/bridge/models/conversion/model_bridge.py]: Our design makes should_quantize a callable so that we can pass the module name to the should_quantize function and it can decide whether we should quantize the corresponding model weight. Usually, weights in linear modules or MoE modules are quantized while weights in layernorm, embedding layers, etc are not. Users can customize which weights to quantize.

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 5]
```diff
@@ -723,6 +724,79 @@ def stream_weights_megatron_to_hf(
...
+                if not merge_adapter_weights and "to_wrap.weight" in task.global_param_name:
+                    hf_name = hf_name[: -len("weight")] + "base_layer.weight"
+
+                if embeddings_are_tied and hf_name == "model.embed_tokens.weight":
+                    yield HFWeightTuple(hf_name, final_tensor)
+                    if hasattr(hf_pretrained, "state") and hasattr(hf_pretrained.state, "source"):
+                        expected_keys = hf_pretrained.state.source.get_all_keys()
+                        if "lm_head.weight" in expected_keys:
+                            yield HFWeightTuple("lm_head.weight", final_tensor.clone().detach())
+                elif embeddings_are_tied and hf_name == "lm_head.weight":
```
[the remark this answers is not in the document]
Comment (hy2826) [reply src/megatron/bridge/models/conversion/model_bridge.py]: Token embedding layers and lm_head are usually not quantized as quantizing those weights will significantly harm the model performance.

File: src/megatron/bridge/models/conversion/model_bridge.py [hunk 6]
```diff
@@ -723,6 +724,79 @@ def stream_weights_megatron_to_hf(
...
+            if merge_adapter_weights and "to_wrap.weight" in task.global_param_name:
+                task_global_base_prefix, _, _ = task.global_param_name.partition(".to_wrap.weight")
+                adapter_tasks = adapter_tasks_by_base.get(task_global_base_prefix)
+            if merge_adapter_weights and adapter_tasks:
+                adapter_weights = self.materialize_adapter_weights(adapter_tasks)
+                converted_weights_dict = self._merge_lora_adapter_weights(
+                    megatron_model,
+                    converted_weights_dict,
+                    adapter_weights,
+                )
```
[the remark this answers is not in the document]
Comment (hy2826) [reply src/megatron/bridge/models/conversion/model_bridge.py]: Currently, our design does not support quantization + LoRA. We will remove the part on LoRA and add an assertion to ensure merge_adapter_weights is False.

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 2 above)
Comment (yaoyu-33) [reply src/megatron/bridge/models/conversion/model_bridge.py]: understand, rename to quantize checker maybe, should quantize sounds like a bool.

**Comment by suiyoubi** [comment]
Hi @hy2826 , kindly follow up here if there is any update to the PR, thanks for the contribution!

**Comment by hy2826** [comment]
> Hi @hy2826 , kindly follow up here if there is any update to the PR, thanks for the contribution!

Hi @suiyoubi , thanks for the reminder! The change to the code has been finished. We are finding the machines to test our changes. Sorry for the wait!

**Comment by suiyoubi** [comment]
Thanks @hy2826 for the update! Feel free to ping us again once it is ready for review. Thanks!

File: src/megatron/bridge/models/conversion/model_bridge.py (hunk 4 above)
Comment (hy2826) [reply src/megatron/bridge/models/conversion/model_bridge.py:2063]: Thanks for the suggestion. We have moved the stream_weights_megatron_to_hf_quant function to a new file called quant_bridge.py mimicing peft bridge.

File: src/megatron/bridge/models/conversion/auto_bridge.py (hunk 1 above)
Comment (hy2826) [reply src/megatron/bridge/models/conversion/auto_bridge.py:17]: We have added unit tests in tests/unit_tests/quantization/test_quant_bridge.py to test the new quantization-related functions in param_mapping.py.

**Comment by hy2826** [comment]
Hi @yaoyu-33 @suiyoubi , we have modified our code according to your suggestions. The changes include 

- fixing several bugs pointed out by coderabbit
- refactoring stream_weights_megatron_to_hf_quant into quant_bridge.py
- renaming should_quantize to quantization_checker
- adding unit tests in tests/unit_tests/quantization/test_quant_bridge.py to test the megatron_to_hf_quant functions in different mappings. 

Let us know if there are any further comments! Thanks!

**Comment by hy2826** [comment]
Hi all, the PR is ready for review. Please let us know if there are any further questions. Thanks!

**Comment by yaoyu-33** [comment]
@hy2826  can you check Claude review, sry. Overall lgtm. Already approved, but want to improve a bit since Claude commented.

File: src/megatron/bridge/models/conversion/param_mapping.py [hunk 7]
```diff
@@ -721,6 +732,73 @@ def gather_from_ep_ranks(
             weights_dict[param_name] = weights_dict[param_name].squeeze()
         return weights_dict
 
+    def gather_from_ep_ranks_scale(
```
[the remark this answers is not in the document]
Comment (hy2826) [reply src/megatron/bridge/models/conversion/param_mapping.py:806]: We agree that the logic of gather_from_ep_ranks_scale is the same as gather_from_ep_ranks. Although the implementation is largely the same under FP8 quantization, we choose to make it a separate function so that special treatments can be implemented for other low precision formats like MXFP4, NVFP4, etc in the future.

File: src/megatron/bridge/models/conversion/param_mapping.py [hunk 8]
```diff
@@ -883,6 +961,58 @@ def megatron_to_hf(
         return {str(self.hf_param): full_weights}
```
[the remark this answers is not in the document]
Comment (hy2826) [reply src/megatron/bridge/models/conversion/param_mapping.py:1035]: This follows from the original implementation of Megatron-Bridge to make CollumnParallelMapping and RowParallelMapping two separate implementations.

**Comment by hy2826** [comment]
Hi @yaoyu-33 , thank you very much for the further feedback! We have revised our PR according to Claude's review.
